modules/dynamic-provisioning-azure-disk-definition.adoc

```diff
@@ -15,29 +15,22 @@ metadata:
-is ignored. If a storage account is not specified, a new storage
-account gets created in the same resource group as the cluster. If you
-are specifying a `storageAccount`, the value for `kind` must be `Dedicated`.
-<2> Azure storage account SKU tier. Default is empty. Note that Premium
-VMs can attach both `Standard_LRS` and `Premium_LRS` disks, Standard VMs
-can only attach `Standard_LRS` disks, Managed VMs can only attach
-managed disks, and unmanaged VMs can only attach unmanaged disks.
-<3> Possible values are `Shared` (default), `Dedicated`, and `Managed`.
+<1> Azure storage account name. This must reside in the same resource group as the cluster. If a storage account is specified, the `location` is ignored. If a storage account is not specified, a new storage account gets created in the same resource group as the cluster. If you are specifying a `storageAccount`, the value for `kind` must be `Dedicated`.
+<2> Azure storage account SKU tier. Default is empty. Note that Premium VMs can attach both `Standard_LRS` and `Premium_LRS` disks, Standard VMs can only attach `Standard_LRS` disks, Managed VMs can only attach managed disks, and unmanaged VMs can only attach unmanaged disks.
+<3> Possible values are `Shared` (default), `Managed`, and `Dedicated`.
 +
-.. If `kind` is set to `Shared`, Azure creates all unmanaged disks in a
-few shared storage accounts in the same resource group as the cluster.
+.. If `kind` is set to `Shared`, Azure creates all unmanaged disks in a few shared storage accounts in the same resource group as the cluster.
 .. If `kind` is set to `Managed`, Azure creates new managed disks.
-.. If `kind` is set to `Dedicated` and a `storageAccount` is specified,
-Azure uses the specified storage account for the new unmanaged disk in
-the same resource group as the cluster. For this to work:
+.. If `kind` is set to `Dedicated` and a `storageAccount` is specified, Azure uses the specified storage account for the new unmanaged disk in the same resource group as the cluster. For this to work:
  * The specified storage account must be in the same region.
- * Azure Cloud Provider must have a write access to the storage account.
-.. If `kind` is set to `Dedicated` and a `storageAccount` is not
-specified, Azure creates a new dedicated storage account for the new
-unmanaged disk in the same resource group as the cluster.
+ * Azure Cloud Provider must have write access to the storage account.
+.. If `kind` is set to `Dedicated` and a `storageAccount` is not specified, Azure creates a new dedicated storage account for the new unmanaged disk in the same resource group as the cluster.
++
+[IMPORTANT]
+====
+Red Hat does not support the use of `kind: Dedicated` in the storage class. {product-title} creates a managed disk for machine OS, or root, disks. However, Azure Disk does not allow the use of both managed and unmanaged disks on a node. Therefore, unmanaged disks cannot be attached to {product-title} nodes.
+====
```

Comment:
Based on [this comment](https://bugzilla.redhat.com/show_bug.cgi?id=1918270#c7) in the BZ.

Reply:
@bobfuru From line 26, seems  the volumes created with `kind: Shared` are the unmanaged disks too.

Reply:
@jsafrane IIUC, @qinpingli is indicating here that RH would not support `kind: Shared` either because Azure creates unmanaged disks in that scenario as well. Is that correct, and if so, do we need to update this note to something like the following?

> Red Hat only supports the use of `kind: Managed` in the storage class...etc.

Reply:
I haven't been able to create volumes with `kind: Shared`, but I don't know if it was mis-configuration on my / OCP side or it really can't be supported in OCP. Unfortunately, I did not save the error message.

@qinpingli do you know if we can support `Shared` volumes?

Reply:
@jsafrane Checked our test cases, looks like no cases to set kind = `Shared` explicitly. 
From the [doc](https://kubernetes.io/docs/concepts/storage/storage-classes/#volume-binding-mode), looks like the default kind is `Shared`, tried with the following storageclass:
$ cat sc-default.yaml 
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: slow
provisioner: kubernetes.io/azure-disk
parameters:
  skuName: Standard_LRS
  location: centralus
  storageAccount: XXXXX
Volume can be created successfully, and be attached successfully.

Tried the following storageclass to set the kind explicitly:
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: slow
provisioner: kubernetes.io/azure-disk
parameters:
  skuName: Standard_LRS
  location: centralus
  storageAccount: XXXXX
  kind: Shared
Volume can not be created, get the following event from PVC:
Failed to provision volume with StorageClass "slow": couldn't find key named key1 in storage account:XXXX keys

seems the default kind is not `Shared`, and ocp doesn't support it?

Reply:
@sferich888 It looks like OCP cannot support `kind:Shared` or `kind:Dedicated` for Azure Disk, so I'm looking at adding a note to explain this. 

Do you have any concerns if we were to add a statement here along the lines of the following?

> **!**
> Red Hat only supports the use of `kind: Managed` in the storage class.
> 
> With `Shared` and `Dedicated`, Azure creates unmanaged disks, while {product-title} creates a managed disk for machine OS (root) disks. But because Azure Disk does not allow the use of both managed and unmanaged disks on a node, unmanaged disks created with `Shared` or `Dedicated` cannot be attached to {product-title} nodes.

Reply:
I checked the upstream docs, 

> When kind is shared, all unmanaged disks are created in a few shared storage accounts in the same resource group as the cluster

This implies that shared disk is unmanaged too and can't work in OCP, with the same explanation as for `Dedicated` disks. The note in the previous comment looks good to me.

Reply:
Thanks for confirming, @jsafrane. 

FYI, this will need to go through our change management process, requiring an ack from @sferich888 or @reestr prior to merge.

Reply:
@davemulford your probably best to review this due to its connection to Azure. 

However as I understand it - were simply clarifying that due to the way OpenShift (in IPI/UPI deployments deploys), customers on Azure are limited to a specific disk type for PV's (when using the Azure Dynamic Storage Provider (as a storage class)). 

From a docs point of view: https://docs.microsoft.com/en-us/azure/virtual-machines/managed-disks-overview and https://docs.microsoft.com/en-us/azure/virtual-machines/disks-types (may be keep partner docs to link to) to help customers understand what it is were talking about. 

If it were me I would ack this, but I think @davemulford should review too.

Reply:
@jsafrane why doen't the product (API's) or operators; enforce this ? IE; why don't we use validations to say; this isn't a legal config on OpenShift?

Reply:
The wording is correct and technically accurate, so the docs change is fine.

I do agree with @sferich888 that we should be validating this in some form.  Whether that's through the product itself, or an operator, doesn't really matter to me.

Reply:
@sferich888, traditionally, StorageClass.Parameters are not validated anywhere in Kubernetes and OpenShift. We could add a webhook / admission plugin, but IMO it could break some user automation scripts / operators - objects previously accepted by API server would now be rejected. Those probably deserve to be broken though.

In addition, it will become endless whack-a-mole when upstream introduces new parameters.